[Pages] functions + get started revision (#2726)
cloudflare/cloudflare-docs: products/pages/src/content/platform/functions.md

 return new Response(`Hello world`);
}
```

### Handling multiple requests in a single function

> You can define multiple HTTP handlers in a single file by exporting the multiple Functions. For example, this file will handle `POST` and `PUT` requests with the same code: 

Author: @lukeed 'the multiple Functions' sounds very awkward, I'm checking if this is intentional because this particular page switches from Functions (product name, capitalised) to functions (noun, lower case) multiple times.

> You can define multiple HTTP handlers in a single file by exporting the multiple Functions. For example, this file will handle `POST` and `PUT` requests with the same code: 

Reviewer: ```suggestion
You can define multiple HTTP handlers in a single file by defining multiple exports within the same file. For example, this file will handle `POST` and `PUT` requests with the same handler code: 
```

```js
export async function onRequestPost(context) {
 // ...
}
export const onRequestPut = onRequestPost;
```

Additionally, an exported handler may be an array of function handlers. This allows you to easily compose Functions as a group, which may include a mix of shared and/or one-off behaviors:
 
```js
import { extraLogging } from 'middlewares.ts'

export const onRequest = [
 extraLogging,
 async ({ request }) => {
   // ...
 }
]
```

## Adding middleware

> Middleware are reusable chunks of logic that can be before and/or after route handlers. These are typically utility Functions that should be applied for chunks of an application’s routes. For example, error handling, user authentication, and logging are typical candidates for middleware within an application.

Author: @lukeed "Middleware are reusable chunks of logic that can be _____ before" I feel like a word was lost here.

> Middleware are reusable chunks of logic that can be before and/or after route handlers. These are typically utility Functions that should be applied for chunks of an application’s routes. For example, error handling, user authentication, and logging are typical candidates for middleware within an application.

Reviewer: ```suggestion
Middleware are reusable chunks of logic that can be executed before and/or after route handlers. These are typically utility Functions that should be applied for chunks of an application’s routes. For example, error handling, user authentication, and logging are typical candidates for middleware within an application.
```

    return new Response(`${err.message}\n${err.stack}`, { status: 500 })
  }
}

// Attach `errorHandler` to all HTTP requests
export const onRequest = errorHandler;
```

### Middleware data

> Handler functions have the ability to pass data between one another. This is done through the `context.data` property, which is accessible and mutable by all functions throughout a request’s execution. 

Author: @lukeed are all the 'functions' here just nouns and not capitalized product names?

Reviewer: "Handler functions" --> functions are not special; theyre just a code function
"Functions" --> the name of the feature, like Workers

> Handler functions have the ability to pass data between one another. This is done through the `context.data` property, which is accessible and mutable by all functions throughout a request’s execution. 

Reviewer: ```suggestion
Handler functions have the ability to pass data between one another. This is done through the `context.data` property, which is accessible and mutable by all handlers throughout a request’s execution. 
```

> More often than not, `context.data` only applies from a middleware’s perspective, but it is available to all functions regardless.

Reviewer 2: `More often than not, `context.data` only applies from a middleware’s perspective` feels awkward. Does it make sense to change it to `More often than not, `context.data` only applies to middleware data`? --> Not sure if `middleware` is data, function, type...

> More often than not, `context.data` only applies from a middleware’s perspective, but it is available to all functions regardless.

Reviewer: ```suggestion
More often than not, `context.data` is only relevant from a middleware’s perspective, but it is available to all functions regardless.
```

### KV namespace

Workers KV is Cloudflare’s globally replicated key-value storage solution. Within Pages, you can choose from the list of KV namespaces that you created from **Account Home** > **Workers** > **KV**. Refer to the [Workers KV documentation for further instructions on adding KV namespace bindings](https://developers.cloudflare.com/workers/platform/environment-variables#adding-kv-namespace-bindings-via-the-dashboard). 

Once your binding is created in the Workers dashboard, go to **your Pages project** > **Settings** > **Functions**. Select the *namespace* from the dropdown and set a name for your binding. 

![KV-Binding](KV-functions.png)

### Durable Object namespace

Durable Objects are Cloudflare’s strongly consistent coordination primitive that power capabilities such as connecting WebSockets, handling state, and building applications. As with Workers KV, you first have to [create the Durable Object](https://developers.cloudflare.com/workers/learning/using-durable-objects#uploading-a-durable-object-worker). You can then configure it as a binding to your Pages project. 

> Go to **Account Home** > **Pages** > **your Pages project** > **Settings** > **Functions** to select the namespace and set a name. 

Author: @lukeed these feel like incomplete instructions because we are introducing DO very abruptly here. After the user "creates a DO" is all that is required of them next to 1) select the namespace and 2) set a name? Either 281 needs more context or 283 does.

Reviewer: i don't know the rest of the instructions. we can come back thru and update after had a chance to try it personally

  }
}
```

## Adding bindings 

While bringing your Workers to Pages, bindings are a big part of what makes your application truly full-stack. You can add KV, Durable Object, and plain-text bindings to your project. 

### KV namespace

> Workers KV is Cloudflare's globally replicated key-value storage solution. Within Pages, you can choose from the list of KV namespaces that you created from **Account Home** > **Workers** > **KV**. Refer to the [Workers KV documentation for further instructions on adding KV namespace bindings](https://developers.cloudflare.com/workers/platform/environment-variables#adding-kv-namespace-bindings-via-the-dashboard). 

Reviewer: ```suggestion
Workers KV is Cloudflare's globally replicated key-value storage solution. Within Pages, you can choose from the list of KV namespaces that you created from **Account Home** > **Pages** > **your Pages project** > **Settings** > **Functions** > **KV namespace bindings**. Select **Add binding** and input a **Variable name** and select a *Kv namespace* from the list of your existing Workers KV namespaces. You will need to repeat this for both the **Production** and **Preview** environments.
```

![KV-Binding](KV-functions.png)

### Durable Object namespace

Durable Objects are Cloudflare's strongly consistent coordination primitive that power capabilities such as connecting WebSockets, handling state, and building applications. As with Workers KV, you first have to [create the Durable Object](https://developers.cloudflare.com/workers/learning/using-durable-objects#uploading-a-durable-object-worker). You can then configure it as a binding to your Pages project. 

> Go to **Account Home** > **Pages** > **your Pages project** > **Settings** > **Functions** to select the namespace and set a name. 

Reviewer: ```suggestion
Go to **Account Home** > **Pages** > **your Pages project** > **Settings** > **Functions** > **Durable Object bindings**. Select **Add binding** and input a **Variable name** and select a *Durable Object namespace* from the list of your existing Durable Objects. You will need to repeat this for both the **Production** and **Preview** environments.
```

![DO-Binding](DO-functions.png)

### Environment variable

> An environment variable is an injected value that can be accessed by your Functions. It is stored as plain-text. You can set your environment variables directly within the Pages interface for both your production and preview environments at run-time and build-time. 

Author: ```suggestion
An [environment variable](https://developers.cloudflare.com/workers/platform/environment-variables) is an injected value that can be accessed by your Functions. It is stored as plain-text. You can set your environment variables directly within the Pages interface for both your production and preview environments at run-time and build-time. 
```

npx wrangler pages dev -- npx react-scripts start
```

Developing locally does not deploy your changes. It is only a means to preview and test. To deploy your changes to your Pages site, you will need to `git commit` and `git push` as normal.

## Pricing and limits

While still in open beta, there is no additional cost to deploy Functions within a Pages project. However, during beta, there is a total limit of 100,000 invocation requests per day. If you reach your daily limit, Pages will stop executing your Functions and fallback to static assets only.

If you reach this limit or need it increased in order to migrate your existing Workers applications into a Pages project, [complete this form](https://forms.gle/fK65trEL67cTvGQG6) to request an increase. You may track this limit by going to **Account Home** > **Pages** > and reviewing **Functions requests today**.

> In the future, you can expect billing to reflect that of the Workers Bundled plan, as Functions are Workers under the hood. 

Reviewer: ```suggestion
In the future, you can expect billing to reflect that of the Workers Bundled plan. 
```
